rotkehlchen/chain/ethereum/modules/uniswap/v3/decoder.py

```diff
@@ -57,93 +69,245 @@ def _maybe_decode_v3_swap(  # pylint: disable=no-self-use
+    ) -> Optional[SwapData]:
+        from_asset: Optional[Asset] = None
+        from_amount = ZERO
+
+        # In this loop we find all the events that spend the token and sum them up.
+        # `0` event is the gas event, `1` event is the ETH receipt event, so we start with `2`.
+        for event in decoded_events[2:]:
+            if (
+                event.event_type == HistoryEventType.INFORMATIONAL and
+                event.event_subtype == HistoryEventSubType.APPROVE
+            ):
+                continue  # skip approval events
+
+            if (
+                event.event_type == HistoryEventType.TRADE and
+                event.event_subtype == HistoryEventSubType.SPEND and
+                event.counterparty in {CPT_UNISWAP_V2, CPT_UNISWAP_V3}
+            ) is False:
+                return None  # unexpected event
+
+            if from_asset is None:
+                from_asset = event.asset
+            elif from_asset != event.asset:
+                return None  # unexpected event
+
+            from_amount += event.balance.amount
+
+        return SwapData(
+            from_asset=from_asset,
+            from_amount=from_amount,
+            to_asset=A_ETH,
+            to_amount=decoded_events[1].balance.amount,  # the eth receipt event
+        )
+
+    def _decode_token_to_token_swap(
+            self,
+            decoded_events: list[HistoryBaseEntry],
+    ) -> Optional[SwapData]:
+        from_asset: Optional[Asset] = None
+        from_amount = ZERO
+        to_asset: Optional[Asset] = None
+        to_amount = ZERO
 
-        The swap method has as data the delta on the pool for the assets swapped so we make sure
-        to use signed integers to detect the amounts.
+        # In this loop we sum up events that spend `from_asset` and sum up
+        # events that receive the `to_asset`. `0` is the gas event, so we start with `1`.
+        for event in decoded_events[1:]:
+            if (
+                event.event_type == HistoryEventType.INFORMATIONAL and
+                event.event_subtype == HistoryEventSubType.APPROVE
+            ):
+                continue  # skip approval events
+
+            if (
+                event.event_type == HistoryEventType.TRADE and
+                event.counterparty in {CPT_UNISWAP_V2, CPT_UNISWAP_V3}
+            ) is False:
+                # In some swaps the asset is sent from the router to the user at the very end,
+                # after all SWAP events and thus is not labeled with the appropriate counterparty.
+                # In this case we check it here.
+                if to_asset is None and event.event_type == HistoryEventType.RECEIVE:
+                    to_asset = event.asset
+                    to_amount = event.balance.amount
+                else:
+                    return None  # unexpected event
+
+            elif event.event_subtype == HistoryEventSubType.SPEND:
+                if from_asset is None:
```

Comment:
why don't you need to check the counterparties as in the previous check?

Reply:
Because they are already checked by the previous check. If counterparty is not in uniswap-v2 and uniswap-v3, then the body of the `if` above will be called

Reply:
> if counterparty is not in uniswap-v2 and uniswap-v3, then the body of the if above will be called

That is a very very very sneaky thing to reason about by just looking at the code. You either need to rewrite the logic to make this a lot more obvious or put small and strategic comments to explain what you are doing here.

I spent way too long a time looking at this trying to understand your reasoning.

Reply:
Rewrote. Hopefully should be more understandable